Add inventory plugin for Stackpath Edge Compute

shayrybak opened the pull request:
##### SUMMARY
This adds an inventory plugin for Stackpath's Edge Compute product.
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
stackpath_compute inventory plugin
##### ADDITIONAL INFORMATION
This leverages the Stackpath public API to get an inventory of hosts for both containers and VM workloads.

Details on Stackpath Edge Compute https://www.stackpath.com/products/edge-compute/
Details on the Stackpath API https://stackpath.dev/reference

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+DOCUMENTATION = '''
+    name: stackpath_compute
+    plugin_type: inventory
+    short_description: StackPath Edge Computing inventory source
```
```suggestion
    short_description: StackPath Edge Computing inventory source
    version_added: 1.2.0
```

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+            description: >
+                A token that ensures this is a source file for the 'stackpath_compute' plugin.
+            required: True
+            choices: ['stackpath_compute']
```
```suggestion
            choices: ['community.general.stackpath_compute']
```

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+
+# example using credentials to fetch all workload instances in a stack.
+---
+plugin: stackpath_compute
```
```suggestion
plugin: community.general.stackpath_compute
```

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+        - Uses a YAML configuration file that ends with stackpath_compute.(yml|yaml).
+    options:
+        plugin:
+            description: >
+                A token that ensures this is a source file for the 'stackpath_compute' plugin.
```
Please use lists (every item is a paragraph) instead of strings:
```suggestion
            description:
                - A token that ensures this is a source file for the 'stackpath_compute' plugin.
```

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+        plugin:
+            description: >
+                A token that ensures this is a source file for the 'stackpath_compute' plugin.
+            required: True
```
```suggestion
            required: true
            type: str
```
Please use lowercase `true`/`false` in YAML. Also, always add the correct type.

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+            description: >
+                An OAuth client ID generated from the API Management section of the StackPath customer portal
+                U(https://control.stackpath.net/api-management)
+            requierd: True
```
```suggestion
            required: true
```

felixfontein reviewed (commented):
I've commented some basic things. Also, please note that all new plugins are required to have tests (integration and/or unit tests).

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+    plugin_type: inventory
+    short_description: StackPath Edge Computing inventory source
+    requirements:
+        - requests
```
```suggestion
        - requests
```
`When fetching URLs, use fetch_url or open_url from ansible.module_utils.urls. Do not use urllib2, which does not natively verify TLS certificates and so is insecure for https.` . it's from https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#module-security.
can't we use the suggested functions here? (i'm not an expert)

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+                A list of Stack IDs to query instnaces in. If no entry then get instances in all stacks on the account
+                U(https://developer.stackpath.com/docs/en/getting-started/#get-your-stack-id)
+            required: False
+        use_internal_ip:
+            description: >
+                Whether or not to use internal IP addresses, yes to use internal addresses, no to use external addresses.
+                Defaults to external
+            requiered: False
+            choices: ['yes', 'no']
```
```suggestion
        use_internal_ip:
            description:
                - Whether or not to use internal IP addresses. If C(no), uses external IP addresses, internal otherwise.
            requiered: false
            type: bool
            default: ...
```
or something

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+- my_first_stack_id
+- my_other_stack_id
+use_internal_ip: no
+
```
```suggestion
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+'''
+
+EXAMPLES = '''
+
+# example using credentials to fetch all workload instances in a stack.
```
```suggestion
# Example using credentials to fetch all workload instances in a stack.
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+        self.client_secret = None
+        self.stack_id = None
+        self.api_host = "https://gateway.stackpath.com"
+        self.group_keys = [
+            "stackId", "workloadId", "cityCode",
+            "countryCode", "continent", "target", "name", "workloadSlug"]
```
```suggestion
        self.group_keys = [
            "stackId",
            "workloadId",
            "cityCode",
            "countryCode",
            "continent",
            "target",
            "name",
            "workloadSlug",
        ]
```
just considered as a good practice

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+            {
+                "client_id": self.client_id,
+                "client_secret": self.client_secret,
+                "grant_type": "client_credentials"
```
```suggestion
                "grant_type": "client_credentials",
```

Andersson007 reviewed (commented):
@shayrybak thanks for the PR!
Left several comments in addition to what @felixfontein mentioned.

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,242 @@
+            required: True
+        stack_ids:
+            description: >
+                A list of Stack IDs to query instnaces in. If no entry then get instances in all stacks on the account
```
```suggestion
                A list of Stack IDs to query instances in. If no entry then get instances in all stacks on the account
```

shayrybak commented:
Thank you for your comments, @Andersson007 and @felixfontein, I'm trying to carve some time out to get this updated, I hope I'll have something this week.

shayrybak commented:
I've addressed all of the comments and added a basic test.

I will add more tests but I am running into an issue when trying to run the tests locally.

I'm getting 

`Shays-MacBook-Pro:general shay$ ansible-test units --docker default -vvv tests/unit/plugins/inventory/test_stackpath_compute.py
RLIMIT_NOFILE: (256, 9223372036854775807)
MAXFD: 256
ERROR: Target pattern not matched: tests/unit/plugins/inventory/test_stackpath_compute.py`

I'm getting the same type of error trying to run any of the existing units in the repository as well, so I believe I have something 
wrong, but the error message is not very helpful here.

Andersson007 commented:
@shayrybak could you please mark the resolved conversation as resolved by "Resolve conversation" button

Andersson007 commented:
Did you run them as following?
```
ansible-test units tests/unit/plugins/inventory/test_stackpath_compute.py --docker -vvv
```

shayrybak commented:
Yes, I ran that command and got the same error message,

```
RLIMIT_NOFILE: (256, 9223372036854775807)
MAXFD: 256
ERROR: Target pattern not matched: tests/unit/plugins/inventory/test_stackpath_compute.py
```

Andersson007 commented:
@shayrybak have you managed to run the tests locally? I'm seeing a lot of related commits

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            required: false
+            type: list
+        use_internal_ip:
+            description: >
+                Whether or not to use internal IP addresses, If false, uses external IP addresses, internal otherwise.
+                If an instance doesn't have an external IP it will not be returned when this option is set to false.
```
```suggestion
            description:
                - Whether or not to use internal IP addresses, If false, uses external IP addresses, internal otherwise.
                - If an instance doesn't have an external IP it will not be returned when this option is set to false.
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            required: true
+            choices: ['community.general.stackpath_compute']
+        client_id:
+            description: >
+                An OAuth client ID generated from the API Management section of the StackPath customer portal
+                U(https://control.stackpath.net/api-management)
```
```suggestion
            description:
                - An OAuth client ID generated from the API Management section of the StackPath customer portal
                  U(https://control.stackpath.net/api-management).
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+    options:
+        plugin:
+            description:
+                - A token that ensures this is a source file for the 'stackpath_compute' plugin.
```
```suggestion
                - A token that ensures this is a source file for the plugin.
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            requierd: true
+            type: string
+        client_secret:
+            description: >
+                An OAuth client secret generated from the API Management section of the StackPath customer portal
+                U(https://control.stackpath.net/api-management)
```
```suggestion
            description:
                - An OAuth client secret generated from the API Management section of the StackPath customer portal
                  U(https://control.stackpath.net/api-management).
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+                An OAuth client ID generated from the API Management section of the StackPath customer portal
+                U(https://control.stackpath.net/api-management)
+            requierd: true
+            type: string
```
```suggestion
            type: str
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            description:
+                - A list of Stack slugs to query instances in. If no entry then get instances in all stacks on the account
+            required: false
+            type: list
```
```suggestion
            type: list
            elements: ...
```
type of elements

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+        stack_slugs:
+            description:
+                - A list of Stack slugs to query instances in. If no entry then get instances in all stacks on the account
+            required: false
```
```suggestion
```
this is a default

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            description: >
+                Whether or not to use internal IP addresses, If false, uses external IP addresses, internal otherwise.
+                If an instance doesn't have an external IP it will not be returned when this option is set to false.
+            requiered: false
```
```suggestion
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+                An OAuth client secret generated from the API Management section of the StackPath customer portal
+                U(https://control.stackpath.net/api-management)
+            required: true
+            type: string
```
```suggestion
            type: str
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,283 @@
+            }
+        )
+        headers = {
+            "Content-Type": "application/json"
```
```suggestion
            "Content-Type": "application/json",
```

Andersson007 commented on tests/unit/plugins/inventory/test_stackpath_compute.py:
```diff
@@ -0,0 +1,200 @@
+            'updatedAt': '2020-07-08T01:00:00.000000Z',
+            'slug': 'test1',
+            'createdAt': '2020-07-08T00:00:00.000000Z',
+            'accountId': 'XXXX'
```
```suggestion
            'accountId': 'XXXX',
```
generally a good practice

shayrybak commented:
I have been able to, but not using the documented way, I had to copy the inventory file to the base ansible inventory location and change the import command, then I was able to run the tests manually with pytest.

However, I've finished writing the tests, and I believe this is ready for another review.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Andrew Klychkov <notifications@github.com>
Sent: Wednesday, September 23, 2020 11:32:13 PM
To: ansible-collections/community.general <community.general@noreply.github.com>
Cc: Shay Rybak <shay.rybak@stackpath.com>; Mention <mention@noreply.github.com>
Subject: Re: [ansible-collections/community.general] Add inventory plugin for Stackpath Edge Compute (#856)


@shayrybak<https://github.com/shayrybak> have you managed to run the tests locally? I'm seeing a lot of related commits

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ansible-collections/community.general/pull/856#issuecomment-698105395>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGA762NHZOW5WKYZE3COS6TSHLDU3ANCNFSM4QTG55AA>.

CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+            choices: ['community.general.stackpath_compute']
+        client_id:
+            description:
+                - An OAuth client ID generated from the API Management section of the StackPath customer portal
+                - U(https://control.stackpath.net/api-management).
```
```suggestion
                - An OAuth client ID generated from the API Management section of the StackPath customer portal
                  (U(https://control.stackpath.net/api-management)).
```

shayrybak replied:
I'm not sure I understand the request here.

all of the links in the repository are in the format I have, and all the descriptions use a list for multiline.

can you elaborate on why this requires changing?

felixfontein replied:
Every list item is a paragraph. The first paragraph is a sentence without a period at its end. The second is just an URL with a period. This looks plainly wrong. It should be one paragraph with one sentence. (This is probably also wrong for the other cases this happens in the doc of this plugin, but that's for another PR to fix...)

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+        client_secret:
+            description:
+                - An OAuth client secret generated from the API Management section of the StackPath customer portal
+                - U(https://control.stackpath.net/api-management).
```
Same here.

felixfontein commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+            description:
+                - An OAuth client ID generated from the API Management section of the StackPath customer portal
+                - U(https://control.stackpath.net/api-management).
+            requierd: true
```
```suggestion
            required: true
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+            if client_id != 32:
+                raise AnsibleError("client_id must be 32 characters long")
+        except KeyError:
+            raise AnsibleError("config missing client_id, a required paramter")
```
```suggestion
            raise AnsibleError("config missing client_id, a required parameter")
```
I'd change this to `option`, it's better according to the doc best practices.

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+            if client_secret != 64:
+                raise AnsibleError("client_secret must be 64 characters long")
+        except KeyError:
+            raise AnsibleError("config missing client_id, a required paramter")
```
```suggestion
            raise AnsibleError("config missing client_id, a required parameter")
```

Andersson007 commented on plugins/inventory/stackpath_compute.py:
```diff
@@ -0,0 +1,281 @@
+                        instance["continent"] = instance["location"]["continent"]
+                        instance["target"] = instance["metadata"]["labels"]["workload.platform.stackpath.net/target-name"]
+                        try:
+                            ip = instance[self.hostname_key]
```
```suggestion
                            ip = instance[self.hostname_key]
```
F841 local variable 'ip' is assigned to but never used

Andersson007 commented:
@felixfontein do you have anything else to add?

Andersson007 commented:
@shayrybak thanks for the new plugin!
@felixfontein thanks for reviewing!

shayrybak commented:
Thank you guys for the reviews and feedback.
